### PR TITLE
Media: add a 'Reverse Order' button to the gallery editor dialog

### LINKED ISSUE #339: Editor: Media: add a "Reverse order" button to the edit gallery screen
Can we add back the "Reverse order" button for images in the create a gallery flow?

screen shot 2015-09-29 at tue sep 29 6 39 53 pm<sup>Seen at https://design5279.wordpress.com/wp-admin/post-new.php using Chrome 45.0.2454.101 on Mac OS X 10.10.5 logged in as designsimply (superadmin) proxied.</sup>

Here's a screencast explaining more about it: [5m3s](https://cloudup.com/cMTEDaxWpaL)
# calypso-gaps

### PR DESCRIPTION
**Pull request by alisterscott**
This PR seeks to add a ‘Reverse Order’ button to the add new gallery dialog (similar to wp-admin)

Fixes #339 

**Outstanding issues:**

1. Needs design review - I think it makes sense to have the button with the other inputs on the right hand side instead of just in the Edit component
2. `warning.js:36 Warning: Failed prop type: Required prop `legend` was not specified in `EditorMediaModalFieldset`` I'm not sure we need a legend for this button, and I don't know how to supress this warning
3. No unit tests for the new functionality - since no unit tests exist for this existing component I'm not sure where to start on this

cc: @designsimply for testing 😊

**Before**

No 'Reverse Order' option

before calypso

**After**

reverse images

**WP-Admin**

wp-admin reorder

### CODE REVIEW

**Comment by alisterscott** [comment]
Thanks to @jordwest for his feedback/input so far

**Comment by folletto** [comment]
> Needs design review - I think it makes sense to have the button with the other inputs on the right hand side instead of just in the Edit component

Agreed, but I think the button should have a different structure / placement.

The priority for that feature is rather minor compared to other controls, so we should move that down. 

The best idea is probably to turn the "Random order" check, which is already a bit odd without label, to a dropdown:

screen shot 2016-10-26 at 11 06 30

**Comment by lancewillett** [comment]
Side comment: I love the numbered photos for testing ordering. 💯

**Comment by lancewillett** [comment]
+1 to Davide's idea of combining with other ordering controls.

File: client/post-editor/media-modal/gallery/fields.jsx [hunk 1]
```diff
@@ -15,15 +16,17 @@ import EditorMediaModalFieldset from '../fieldset';
 import SelectDropdown from 'components/select-dropdown';
 import SelectDropdownItem from 'components/select-dropdown/item';
 import FormCheckbox from 'components/forms/form-checkbox';
+import Button from 'components/button';
 import { GalleryColumnedTypes, GallerySizeableTypes } from 'lib/media/constants';
 
-export default React.createClass( {
+export const EditorMediaModalGalleryFields = React.createClass( {
 	displayName: 'EditorMediaModalGalleryFields',
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/fields.jsx]: When assigning `React.createClass` to a variable, you can remove `displayName`, as it's inferred automatically via [`babel-plugin-react-display-name`](https://www.npmjs.com/package/babel-plugin-react-display-name).

File: client/post-editor/media-modal/gallery/fields.jsx [hunk 2]
```diff
@@ -95,6 +98,10 @@ export default React.createClass( {
 		this.props.onUpdateSetting( 'orderBy', event.target.checked ? 'rand' : null );
 	},
 
+	updateReverseOrder() {
+		this.props.onReverse();
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/fields.jsx]: I'd suggest either:

- Checking that the `onReverse` prop exists before calling on it
- Providing a `noop` (either `() => {}` or [Lodash's `_.noop`](https://lodash.com/docs/4.16.4#noop) as a default prop
- Marking the prop as `.isRequired` in `propTypes`

My preference is toward the second

File: client/post-editor/media-modal/gallery/fields.jsx [hunk 3]
```diff
@@ -144,13 +151,26 @@ export default React.createClass( {
 		);
 	},
 
+	renderReverseOrderButton() {
+		const { translate } = this.props;
+
+		return (
+			<EditorMediaModalFieldset>
+				<Button onClick={ this.updateReverseOrder } disabled={ this.props.settings.orderBy === 'rand' } >
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/fields.jsx]: Do you need `updateReverseOrder` as a function on the instance, or could you just pass `this.props.onReverse` directly? This would also help avoid the need for any checking of truthiness on the prop, as `<Button />` is responsible for handling this.

File: client/post-editor/media-modal/gallery/fields.jsx [hunk 4]
```diff
@@ -144,13 +151,26 @@ export default React.createClass( {
...
+	},
+
 	render() {
 		const types = this.getTypeOptions();
 		const links = this.getLinkOptions();
 		const sizes = this.getSizeOptions();
 
 		return (
 			<div className="editor-media-modal-gallery__fields">
+				{ this.renderReverseOrderButton() }
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/fields.jsx]: I've started encouraging more rendering inline in the `render` function when possible, as it keeps all render logic colocated. Here, since `renderReverseOrderButton` is simply returning another element, it would be straightforward to move that element in place of the function call here directly. Only in cases where there's conditional render logic does it become more complicated.

File: client/post-editor/media-modal/gallery/fields.jsx [hunk 5]
```diff
@@ -144,13 +151,26 @@ export default React.createClass( {
...
 				{ this.renderDropdown( this.translate( 'Layout' ), types, 'type' ) }
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/fields.jsx]: I might recommend that you consider updating `this.translate` to `this.props.translate` in the same pull request if you're introducing `localize`.

File: client/post-editor/media-modal/gallery/index.jsx [hunk 6]
```diff
@@ -114,6 +114,13 @@ const EditorMediaModalGallery = React.createClass( {
 		this.props.onUpdateSettings( updatedSettings );
 	},
 
+	reverseOrder( ) {
+		const reverseItems = reverse( this.props.settings.items.slice( 0 ) );
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/index.jsx]: Nice catch on the [mutation note on `_.reverse`](https://lodash.com/docs/4.16.4#reverse) and the need to create a clone of the array 👍 

You could alternatively use newer ES6 semantics to make it more condensed:

```js
const reverseItems = reverse( [ ...this.props.settings.items ] );
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator

File: client/post-editor/media-modal/gallery/index.jsx [hunk 7]
```diff
@@ -136,7 +143,8 @@ const EditorMediaModalGallery = React.createClass( {
 							site={ site }
 							settings={ settings }
 							onUpdateSetting={ this.updateSetting }
-							numberOfItems={ items.length } />
+							numberOfItems={ items.length }
+							onReverse = { this.reverseOrder } />
```
Comment (aduth) [inline client/post-editor/media-modal/gallery/index.jsx]: Extra spaces around the `=` should be removed.

**Comment by alisterscott** [comment]
Thanks for your feedback @folletto!

> The priority for that feature is rather minor compared to other controls, so we should move that down.

Sure, moved to the bottom now:

screen shot 2016-10-27 at 9 54 19 am

> The best idea is probably to turn the "Random order" check, which is already a bit odd without label, to a dropdown

I had a chat to @lancewillett about this and it raises a whole lot of additional questions, like what is 'normal' order - is this the 'initial' order or the uploaded order, and what happens when you manually reorder images then choose normal or reverse. To keep this PR small I'd raised this idea as a separate issue which we can work through some ideas/designs for this to implement separately to this. See: #8967

File: client/post-editor/media-modal/gallery/fields.jsx (hunk 1 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/fields.jsx]: Thanks, I was wondering if we needed that anymore. Removed now.

File: client/post-editor/media-modal/gallery/fields.jsx (hunk 2 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/fields.jsx]: > Providing a noop (either () => {} or Lodash's _.noop as a default prop

Thanks, I did this.

File: client/post-editor/media-modal/gallery/fields.jsx (hunk 3 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/fields.jsx]: Thanks, that makes sense: it's not needed.

File: client/post-editor/media-modal/gallery/fields.jsx (hunk 5 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/fields.jsx]: I thought about that but I didn't want to make this PR too big. Is it ok if I do a separate PR immediately following this to refactor these?

File: client/post-editor/media-modal/gallery/index.jsx (hunk 6 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/index.jsx]: Thanks for the tip - updated 😊

File: client/post-editor/media-modal/gallery/index.jsx (hunk 7 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/index.jsx]: Thanks, removed

File: client/post-editor/media-modal/gallery/fields.jsx (hunk 4 above)
Comment (alisterscott) [inline client/post-editor/media-modal/gallery/fields.jsx]: Thanks, moved this content into render

**Comment by alisterscott** [comment]
@aduth: thanks for your review. I've addressed all the issues (besides this this.props.translate which I was planning to do as a separate PR straight after this one) - so would you mind re-reviewing this? Thanks

**Comment by folletto** [comment]
> I had a chat to @lancewillett about this and it raises a whole lot of additional questions, like what is 'normal' order - is this the 'initial' order or the uploaded order, and what happens when you manually reorder images then choose normal or reverse. To keep this PR small I'd raised this idea as a separate issue which we can work through some ideas/designs for this to implement separately to this.

I don't think that's an issue, because the same question can be raised with the button. In other words: if the issue isn't there with the button, it shouldn't be there with the dropdown. If it's a reason to not make the dropdown, then it's a reason to not do the button either. That sidebar is becoming very crowded.

Ok, I'll move the discussion on the other thread.

**Comment by alisterscott** [comment]
> but the "let's not merge this" is even more relevant after that.

Closing.

**Comment by lancewillett** [comment]
@folletto What amount of weight do you assign this being a parity with wp-admin? That's part of the reason we split out the small change from the bigger discussion and further improvement? It's <a href="https://github.com/Automattic/wp-calypso/milestone/129">a big list (19 open today)</a> and this change would reduce that disconnect a bit, on its own.

I think we should push the change with just the button for parity, then explore the better UX as a second PR.

**Comment by folletto** [comment]
It's not about weight parity, is about adding the button in the wrong part of the UI, a UI that is already full of elements. Adding features and _then_ doing a design for them is the wrong process. Any element on that list should have a design or a design review before it gets developed.

On this specific issue, my design suggestion is in #8967 and the following comment adds on the approach we should take. As you can see there, "feature parity" might not be the full answer to the real issue.
